x-pack/qa/security-tools-tests/src/test/java/org/elasticsearch/xpack/security/authc/service/FileTokensToolTests.java

```diff
@@ -164,6 +185,72 @@ public void testCreateTokenWithInvalidServiceAccount() throws Exception {
+            () -> execute("delete", pathHomeParameter,
+                randomFrom("elastic/foo", "foo/fleet-server", randomAlphaOfLengthBetween(6, 16)),
+                randomAlphaOfLengthBetween(3, 8)));
+        assertThat(e1.getMessage(), containsString("Unknown service account principal: "));
+        assertThat(e1.getMessage(), containsString("Must be one of "));
+
+        // Invalid toke name
+        final String tokenName2 = ServiceAccountTokenTests.randomInvalidTokenName();
+        final String[] args = tokenName2.startsWith("-") ?
+            new String[] { "delete", pathHomeParameter, "elastic/fleet-server", "--", tokenName2 } :
+            new String[] { "delete", pathHomeParameter, "elastic/fleet-server", tokenName2 };
+        final UserException e2 = expectThrows(UserException.class, () -> execute(args));
+        assertThat(e2.getMessage(), containsString(ServiceAccountToken.INVALID_TOKEN_NAME_MESSAGE));
+
+        // Non-exist token
+        final String tokenName3 = randomAlphaOfLengthBetween(3, 8);
+        final UserException e3 = expectThrows(UserException.class,
+            () -> execute("delete", pathHomeParameter, "elastic/fleet-server", tokenName3));
+        assertThat(e3.getMessage(), containsString("Service token [elastic/fleet-server/" + tokenName3 + "] does not exist"));
+    }
+
+    public void testListTokens() throws Exception {
+        execute("list", pathHomeParameter);
+        final String output = terminal.getOutput();
+        assertThat(output, containsString("elastic/fleet-server/server_1\n" +
+            "elastic/fleet-server/server_2\n" +
+            "elastic/fleet-server/server_3"));
+    }
+
+    public void testListTokensByPrincipal() throws Exception {
+        execute("list", pathHomeParameter, "elastic/fleet-server");
+        final String output = terminal.getOutput();
+        assertThat(output, containsString("elastic/fleet-server/server_1\n" +
+            "elastic/fleet-server/server_2\n" +
+            "elastic/fleet-server/server_3"));
+    }
+
+    public void testListTokensNonExist() throws Exception {
+        // Invalid principal
+        final UserException e1 = expectThrows(UserException.class,
+            () -> execute("list", pathHomeParameter,
+                randomFrom("elastic/foo", "foo/fleet-server", randomAlphaOfLengthBetween(6, 16))));
+        assertThat(e1.getMessage(), containsString("Unknown service account principal: "));
+        assertThat(e1.getMessage(), containsString("Must be one of "));
+
+        // Delete all tokens
+        execute("delete", pathHomeParameter, "elastic/fleet-server", "server_1");
+        execute("delete", pathHomeParameter, "elastic/fleet-server", "server_2");
+        execute("delete", pathHomeParameter, "elastic/fleet-server", "server_3");
+        execute("list", pathHomeParameter, "elastic/fleet-server");
+        final String output = terminal.getOutput();
+        assertThat(output, is(emptyString()));
+    }
+
```

Comment:
Can we have a test for when the file itself doesn't exist?

For `list`, that will work fine, because `FileServiceAccountsTokenStore.parseFile` returns an empty map when the file is missing.
For `delete` it will also work fine, but only because we don't write to the file when `map.remove` returns null. Without that, `delete` on a missing file would create the file, which probably isn't what we want.

Reply:
Good point! I added the test for both list and delete.